src/globus_cli/utils.py

```diff
@@ -284,9 +284,68 @@ def resolve_principal_urn(
         raise NotImplementedError("unrecognized principal_type")
 
 
-def _is_uuid(s: str) -> bool:
+def is_uuid(s: str) -> bool:
     try:
         uuid.UUID(s)
         return True
     except ValueError:
         return False
+
+
+K = t.TypeVar("K")
+V = t.TypeVar("V")
+
+
+class LazyDict(t.Dict[K, V]):
```

Comment:
One of the things I'd like is for us to have a tighter signature on `__init__` (so that we can describe valid vs invalid arguments). I think the easiest path to that is to _contain_ a dict rather than _inheriting from_ a dict.

This is not a complete change suggestion, but what about using `typing.Mapping`?

```suggestion
class LazyDict(t.Mapping[K, V]):
```

`__init__` and other elements would also need refinement.

Reply:
I'm inheriting directly from dict rather than mapping because of [json encoding](https://docs.python.org/3/library/json.html#encoders-and-decoders). Unless this thing is a dict, the builtin `json` module can't serialize it without a custom encoder.


```
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python-ce/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "/Users/derekschlabach/.pyenv/versions/3.10.13/lib/python3.10/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/Users/derekschlabach/.pyenv/versions/3.10.13/lib/python3.10/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/Users/derekschlabach/.pyenv/versions/3.10.13/lib/python3.10/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/Users/derekschlabach/.pyenv/versions/3.10.13/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type MyMapping is not JSON serializable
```

This object needs to be serializable by our JsonPrinter for the `-F json` flag & I didn't want to add a custom serializer just for it.

Reply:
Looking at this again, I think we should be using `collections.abc.Mapping`.
The typing generic forms are not really where we want to inherit from for runtime behaviors, and as of 3.9+ the collections classes support class getitem.

Reply:
Try telling that to py38. I had `dict[K, V]` initially but had to [change it](https://github.com/globus/globus-cli/pull/1077/commits/9c8e230f5a693609e7a2d8e47c6d79f3df6100e1) even with `__future__.annotations`

Reply:
This isn't critical, but inheriting from a `typing` definition feels wrong to me, and I'd like to be in a position where we never have to think about it.
Maybe we just punt on the issue.

More thoughts below:

Python 3.8 is out of support. We're still testing against it, but I think we should fix that so that we can write ergonomic code.

I still think the right _runtime_ type to inherit is `dict`. The 3.8-compatible way would be
```python
if t.TYPE_CHECKING:
    LazyDictBase = t.Dict[K, V]
else:
    LazyDictBase = dict
```

But that's pretty unpleasant to handle.

Reply:
I basically agree with everything you've said here (including that being a pretty unpleasant to handle definition).

What if I added a comment to change the inherited class to `dict` once py38 was deprecated?